ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -260,4 +452,359 @@ private void updateMapHelper(Map<String, Long> map, String key, Long value) {
+          // is started and it's going to process new entries with offset 90 to 120. Since IndexScanner is still in
+          // progress the QueueProcessor will operate on the old ScanResults and index entries between 100 and
+          // 120 will be missing for the next forecast period.
+          if (isScanning || currentScanResults == null) {
+            return;
+          } else {
+            entryCount = queueEntryCount.get();
+          }
+        }
+        for (int i = 0; i < entryCount; i++) {
+          if (cancelled) {
+            return;
+          }
+          Pair<IndexValue, IndexValue> entry = recentEntryQueue.poll();
+          if (entry == null) {
+            throw new IllegalStateException(
+                "Invalid queue state. Expected entryCount " + entryCount + " current index: " + i);
+          }
+          queueEntryCount.decrementAndGet();
+          IndexValue newValue = entry.getFirst();
+          // prevent double counting new entries that were added after enabling the queue and just before the second
+          // checkpoint is taken
+          if (newValue.getOffset().compareTo(currentScanResults.scannedEndOffset) >= 0) {
+            if (newValue.isFlagSet(IndexValue.Flags.Delete_Index)) {
+              // new delete
+              processNewDelete(currentScanResults, newValue, entry.getSecond());
+            } else {
+              // new put
+              processNewPut(currentScanResults, newValue);
+            }
+          }
+        }
+      } catch (Exception e) {
+        logger.error("Unexpected exception while running QueueProcessor", e);
+        metrics.blobStoreStatsErrorCount.inc();
+      }
+    }
+
+    void cancel() {
+      cancelled = true;
+    }
+  }
+
+  /**
+   * Runner that scans the entire index and populate a new {@link ScanResults} to start/extend the forecast coverage so
+   * new requests can be answered using the created {@link ScanResults} instead of another scan of the index.
+   */
+  private class IndexScanner implements Runnable {
+    private volatile boolean cancelled = false;
+    volatile ScanResults newScanResults;
+    long startTimeInMs;
+
+    @Override
+    public void run() {
+      try {
+        recentEntryQueueEnabled = false;
+        startTimeInMs = time.milliseconds();
+        newScanResults = new ScanResults(startTimeInMs, logSegmentForecastOffsetMs, bucketCount, bucketSpanInMs);
+        synchronized (scanLock) {
+          isScanning = true;
+        }
+        Offset firstCheckpoint = index.getCurrentEndOffset();
+        ConcurrentNavigableMap<Offset, IndexSegment> indexSegments = index.getIndexSegments();
+        Map<StoreKey, Long> deletedKeys = new HashMap<>();
+        // process the active index segment based on the firstCheckpoint in case index segment rolled over after the
+        // checkpoint is taken
+        Offset activeIndexSegmentStartOffset = indexSegments.floorKey(firstCheckpoint);
```

Comment:
this can be `null` if there are no entries at all - please add a test for this case.